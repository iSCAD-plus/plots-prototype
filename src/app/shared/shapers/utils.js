import R from 'ramda';
import S from 'sanctuary';

// debug use only
const log = R.tap(console.log);

const shapeMultiSeries = R.compose(
  R.map(
    R.compose(
      R.zipObj(['key', 'values']),
      R.over(R.lensIndex(1), R.map(R.omit('seriesKey'))),
    ),
  ),
  R.toPairs,
  R.groupBy(R.prop('seriesKey')),
);

const shapeSingleSeries = R.compose(
  R.merge({key: 'series1'}),
  R.objOf('values'),
);

const isMultiSeries = R.compose(
  S.isJust,
  S.chain(S.get(S.is(String), 'seriesKey')),
  S.head,
);

export const getBarChartValues = R.ifElse(
  isMultiSeries,
  shapeMultiSeries,
  shapeSingleSeries,
);
