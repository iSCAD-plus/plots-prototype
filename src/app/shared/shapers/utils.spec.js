import { getBarChartValues } from './utils';

const dataPoint = { xKey: 1, yKey: 'axisName' };

const multiSeries = [
  { ...dataPoint, seriesKey: 'series1' },
  { ...dataPoint, seriesKey: 'series2' },
];

const singleSeries = [
  dataPoint,
  dataPoint,
];

describe('#getBarChartValues', () => {
  describe('should produce correct shape when seriesKey is:', () => {
    it('present', () => {
      expect(getBarChartValues(multiSeries)).toMatchSnapshot();
    });

    it('absent', () => {
      expect(getBarChartValues(singleSeries)).toMatchSnapshot();
    });
  });
});
