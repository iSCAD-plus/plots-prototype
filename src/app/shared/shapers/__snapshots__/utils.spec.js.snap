// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#getBarChartValues should produce correct shape when seriesKey is: absent 1`] = `
Object {
  "key": "series1",
  "values": Array [
    Object {
      "xKey": 1,
      "yKey": "axisName",
    },
    Object {
      "xKey": 1,
      "yKey": "axisName",
    },
  ],
}
`;

exports[`#getBarChartValues should produce correct shape when seriesKey is: present 1`] = `
Array [
  Object {
    "key": "series1",
    "values": Array [
      Object {
        "xKey": 1,
        "yKey": "axisName",
      },
    ],
  },
  Object {
    "key": "series2",
    "values": Array [
      Object {
        "xKey": 1,
        "yKey": "axisName",
      },
    ],
  },
]
`;
